assets/js/query/lex.ts

```diff
@@ -0,0 +1,192 @@
+          }
+          else {
+            negate = !negate;
+          }
+          break;
+        case 'lparen':
+          if (searchTerm) {
+            // If we are inside the search term, do not error out just yet;
+            // instead, consider it as part of the search term, as a user convenience.
+            searchTerm += token;
+            lparenCtr += 1;
+          }
+          else {
+            opQueue.unshift('lparen');
+            groupNegate.push(negate);
+            negate = false;
+          }
+          break;
+        case 'rparen':
+          if (lparenCtr > 0) {
+            searchTerm = assertNotNull(searchTerm) + token;
+            lparenCtr -= 1;
+          }
+          else {
+            while (opQueue.length > 0) {
+              const op = assertNotUndefined(opQueue.shift());
+              if (op === 'lparen') {
+                break;
+              }
+              tokenStack.push(op);
+            }
+            if (groupNegate.length > 0 && groupNegate.pop()) {
+              tokenStack.push('not_op');
+            }
+          }
+          break;
+        case 'fuzz':
+          if (searchTerm) {
+            // For this and boost operations, we store the current match so far
+            // to a temporary string in case this is actually inside the term.
+            fuzz = parseFloat(token.substring(1));
+            boostFuzzStr += token;
+          }
+          else {
+            searchTerm = token;
+          }
+          break;
+        case 'boost':
+          if (searchTerm) {
+            boost = parseFloat(token.substring(1));
+            boostFuzzStr += token;
+          }
+          else {
+            searchTerm = token;
+          }
+          break;
+        case 'quoted_lit':
+          if (searchTerm) {
+            searchTerm += token;
+          }
+          else {
+            searchTerm = token;
+          }
+          break;
+        case 'word':
+          if (searchTerm) {
+            if (fuzz !== 0 || boost !== 1) {
+              boost = 1;
+              fuzz = 0;
+              searchTerm += boostFuzzStr;
+              boostFuzzStr = '';
+            }
+            searchTerm += token;
+          }
+          else {
+            searchTerm = token;
+          }
+          break;
+        default:
+          // Append extra spaces within search terms.
+          if (searchTerm) {
+            searchTerm += token;
+          }
+      }
+
+      // Truncate string and restart the token tests.
+      localSearchStr = localSearchStr.substring(token.length);
+
+      // Break since we have found a match.
+      break;
+    }
+  }
+
+  // Append final tokens to the stack.
+  pushTerm();
+
+  if (opQueue.indexOf('rparen') !== -1 || opQueue.indexOf('lparen') !== -1) {
+    throw new Error('Mismatched parentheses.');
+  }
+
+  // Memory-efficient concatenation of remaining operators queue to the
+  // token stack.
+  tokenStack.push.apply(tokenStack, opQueue);
```

Comment:
This is the kind of thing that IMO feels like micro-optimization in JS and makes the code less readable. From a code clarity standpoint `tokenStack.push(...opQueue)` is more straight-forward. I leave it up to you.